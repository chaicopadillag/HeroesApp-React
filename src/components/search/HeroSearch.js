import React, { useMemo } from 'react';
import queryString from 'query-string';
import useForm from '../../hooks/useform';
import HeroCard from '../heroes/HeroCard';
import { useLocation } from 'react-router-dom';
import getHeroesByName from '../../selectors/getHeroesByName';

const HeroSearch = ({ history }) => {
	const location = useLocation(),
		{ keyword = '' } = queryString.parse(location.search),
		[formValues, handleChange] = useForm({ search: keyword }),
		{ search } = formValues;
	const heroesResult = useMemo(() => getHeroesByName(keyword), [keyword]);
	const handleSearch = (e) => {
		e.preventDefault();
		history.push(`?keyword=${search}`);
	};
	return (
		<div className="container mt-2">
			<h2 className="text-center">Search Heroes</h2>
			<hr />
			<div className="row">
				<div className="col-6 offset-3">
					<form method="POST" onSubmit={handleSearch} className="row d-flex justify-content-between">
						<div className="col-8">
							<input type="text" className="form-control" placeholder="Keywork" name="search" value={search} onChange={handleChange} autoComplete="off" />
						</div>
						<div className="col-4">
							<button className="btn btn-outline-primary">Search</button>
						</div>
					</form>
				</div>
				<div className="col-12">
					<h2>Resul search</h2>
					<div className="row">
						{keyword === '' && <div className="alert alert-success">Search a Hero</div>}
						{keyword !== '' && heroesResult.length <= 0 && <div className="alert alert-danger">Keyword into not found in database</div>}
						{heroesResult.map((hero) => (
							<HeroCard key={hero.id} {...hero} />
						))}
					</div>
				</div>
			</div>
		</div>
	);
};

export default HeroSearch;
