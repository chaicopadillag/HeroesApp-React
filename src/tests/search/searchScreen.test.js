import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import HeroSearch from '../../components/search/HeroSearch';

describe('Prueba a component <SearchScreen/>', () => {
	test('debe mostrar los valores por defecto', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search']}>
				<Route path="/search" component={HeroSearch} />
			</MemoryRouter>
		);
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('.alert-success').text().trim()).toBe('Search a Hero');
	});
	test('debe de mostrar un hero: Batman con el valor de queryString', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?keyword=Batman']}>
				<Route path="/search" component={HeroSearch} />
			</MemoryRouter>
		);
		expect(wrapper.find('input').prop('value')).toBe('Batman');
		expect(wrapper.find('HeroCard').exists()).toBe(true);
	});

	test('debe mostrar un error si no se encuentra hero', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?keyword=dfg45dgd5g45d5454545']}>
				<Route path="/search" component={HeroSearch} />
			</MemoryRouter>
		);
		expect(wrapper.find('.alert-danger').text().trim()).toBe('Keyword into not found in database');
		expect(wrapper.find('HeroCard').exists()).toBe(false);
	});

	test('Debe de enviar el Formulario', () => {
		const history = {
			push: jest.fn(),
		};
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?keyword=batman']}>
				<Route path="/search" component={() => <HeroSearch history={history} />} />
			</MemoryRouter>
		);
		wrapper.find('input').simulate('change', {
			target: { name: 'search', value: 'batman' },
		});
		wrapper.find('form').prop('onSubmit')({ preventDefault() {} });
		expect(history.push).toHaveBeenCalledWith('?keyword=batman');
	});
});
