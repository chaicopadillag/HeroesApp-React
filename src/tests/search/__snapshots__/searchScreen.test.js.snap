// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba a component <SearchScreen/> debe mostrar los valores por defecto 1`] = `
<div
  className="container mt-2"
>
  <h2
    className="text-center"
  >
    Search Heroes
  </h2>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-6 offset-3"
    >
      <form
        className="row d-flex justify-content-between"
        method="POST"
        onSubmit={[Function]}
      >
        <div
          className="col-8"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="search"
            onChange={[Function]}
            placeholder="Keywork"
            type="text"
            value=""
          />
        </div>
        <div
          className="col-4"
        >
          <button
            className="btn btn-outline-primary"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-12"
    >
      <h2>
        Resul search
      </h2>
      <div
        className="row"
      >
        <div
          className="alert alert-success"
        >
          Search a Hero
        </div>
      </div>
    </div>
  </div>
</div>
`;
